In [324]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [387]:
data = pd.read_csv('weatherHistory.csv')

In [388]:
data.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [327]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


In [328]:
data.shape

(96453, 12)

In [329]:
data.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
count       96453.000000            96453.000000     96453.000000     96453.0   
mean           10.810640              187.509232        10.347325         0.0   
std             6.913571              107.383428         4.192123         0.0   
min             0.000000                0.000000         0.000000         0.0   
25%             5.828200              116.000000         8.339800         0.0   
50%             9.965900              180.000000        10.046400         0.0   
75%            14.135800              290.000000        14.812000         0.0   
max            63.852600              359.000000        16.100000         0.0   

       Pressure (millibars)  
count          96453.000000  
mean            1003.235956  
std              116.969906  
min                0.000000  
25%             1011.900000  
50%             1016.450000  
75%             1021.090000  
max             1046.380000

In [330]:
data.isnull().sum()

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
dtype: int64

#### Dataset has 12 columns and 96453 rows.
#### Precip Type has null values.

In [331]:
def numerical_feature(data):
    numerical_features = [feature for feature in data.columns if data[feature].dtypes != 'O']
    return numerical_features

In [332]:
def categorical_feature(data):
    categorical_feature = [feature for feature in data.columns if data[feature].dtypes == 'O']
    return categorical_feature

In [333]:
categorical_feature(data)

['Formatted Date', 'Summary', 'Precip Type', 'Daily Summary']

In [334]:
numerical_feature(data)

['Temperature (C)',
 'Apparent Temperature (C)',
 'Humidity',
 'Wind Speed (km/h)',
 'Wind Bearing (degrees)',
 'Visibility (km)',
 'Loud Cover',
 'Pressure (millibars)']

#### There are four categorical features and rest are numerical features.

In [335]:
cat = []
ln = []
def get_categories(data):
    for category in categorical_feature(data):
        cat.append(data[category].unique())
        ln.append(len(data[category].unique())) 
    return cat,ln

In [336]:
get_categories(data)

([array(['2006-04-01 00:00:00.000 +0200', '2006-04-01 01:00:00.000 +0200',
         '2006-04-01 02:00:00.000 +0200', ...,
         '2016-09-09 21:00:00.000 +0200', '2016-09-09 22:00:00.000 +0200',
         '2016-09-09 23:00:00.000 +0200'], dtype=object),
  array(['Partly Cloudy', 'Mostly Cloudy', 'Overcast', 'Foggy',
         'Breezy and Mostly Cloudy', 'Clear', 'Breezy and Partly Cloudy',
         'Breezy and Overcast', 'Humid and Mostly Cloudy',
         'Humid and Partly Cloudy', 'Windy and Foggy', 'Windy and Overcast',
         'Breezy and Foggy', 'Windy and Partly Cloudy', 'Breezy',
         'Dry and Partly Cloudy', 'Windy and Mostly Cloudy',
         'Dangerously Windy and Partly Cloudy', 'Dry', 'Windy',
         'Humid and Overcast', 'Light Rain', 'Drizzle', 'Windy and Dry',
         'Dry and Mostly Cloudy', 'Breezy and Dry', 'Rain'], dtype=object),
  array(['rain', 'snow', nan], dtype=object),
  array(['Partly cloudy throughout the day.',
         'Mostly cloudy throughout the 

In [337]:
data['Precip Type'] = data['Precip Type'].fillna(data['Precip Type'].mode)

In [338]:
data

Formatted Date        Summary Precip Type  \
0      2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain   
1      2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain   
2      2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain   
3      2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain   
4      2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain   
...                              ...            ...         ...   
96448  2016-09-09 19:00:00.000 +0200  Partly Cloudy        rain   
96449  2016-09-09 20:00:00.000 +0200  Partly Cloudy        rain   
96450  2016-09-09 21:00:00.000 +0200  Partly Cloudy        rain   
96451  2016-09-09 22:00:00.000 +0200  Partly Cloudy        rain   
96452  2016-09-09 23:00:00.000 +0200  Partly Cloudy        rain   

       Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0             9.472222                  7.388889      0.89            14.1197   
1             9.355556                  7.227778      0.86            14.2646   
2             9.377778                  9.377778      0.89             3.9284   
3             8.288889                  5.944444      0.83            14.1036   
4             8.755556                  6.977778      0.83            11.0446   
...                ...                       ...       ...                ...   
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
0                       251.0          15.8263         0.0   
1                       259.0          15.8263         0.0   
2                       204.0          14.9569         0.0   
3                       269.0          15.8263         0.0   
4                       259.0          15.8263         0.0   
...                       ...              ...         ...   
96448                    31.0          16.1000         0.0   
96449                    20.0          15.5526         0.0   
96450                    30.0          16.1000         0.0   
96451                    20.0          16.1000         0.0   
96452                    39.0          15.5204         0.0   

       Pressure (millibars)                           Daily Summary  
0                   1015.13       Partly cloudy throughout the day.  
1                   1015.63       Partly cloudy throughout the day.  
2                   1015.94       Partly cloudy throughout the day.  
3                   1016.41       Partly cloudy throughout the day.  
4                   1016.51       Partly cloudy throughout the day.  
...                     ...                                     ...  
96448               1014.36  Partly cloudy starting in the morning.  
96449               1015.16  Partly cloudy starting in the morning.  
96450               1015.66  Partly cloudy starting in the morning.  
96451               1015.95  Partly cloudy starting in the morning.  
96452               1016.16  Partly cloudy starting in the morning.  

[96453 rows x 12 columns]

In [339]:
data.drop(['Formatted Date','Summary','Daily Summary'], axis = 1, inplace = True)

In [340]:
data

Precip Type  Temperature (C)  Apparent Temperature (C)  Humidity  \
0            rain         9.472222                  7.388889      0.89   
1            rain         9.355556                  7.227778      0.86   
2            rain         9.377778                  9.377778      0.89   
3            rain         8.288889                  5.944444      0.83   
4            rain         8.755556                  6.977778      0.83   
...           ...              ...                       ...       ...   
96448        rain        26.016667                 26.016667      0.43   
96449        rain        24.583333                 24.583333      0.48   
96450        rain        22.038889                 22.038889      0.56   
96451        rain        21.522222                 21.522222      0.60   
96452        rain        20.438889                 20.438889      0.61   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
0                14.1197                   251.0          15.8263         0.0   
1                14.2646                   259.0          15.8263         0.0   
2                 3.9284                   204.0          14.9569         0.0   
3                14.1036                   269.0          15.8263         0.0   
4                11.0446                   259.0          15.8263         0.0   
...                  ...                     ...              ...         ...   
96448            10.9963                    31.0          16.1000         0.0   
96449            10.0947                    20.0          15.5526         0.0   
96450             8.9838                    30.0          16.1000         0.0   
96451            10.5294                    20.0          16.1000         0.0   
96452             5.8765                    39.0          15.5204         0.0   

       Pressure (millibars)  
0                   1015.13  
1                   1015.63  
2                   1015.94  
3                   1016.41  
4                   1016.51  
...                     ...  
96448               1014.36  
96449               1015.16  
96450               1015.66  
96451               1015.95  
96452               1016.16  

[96453 rows x 9 columns]

#### We don't need date, summary, and daily summary for prediction. These informatio has nothing to do with prediction.

### handling categorical data

In [341]:
obj = data[['Precip Type']]
other = data[['Temperature (C)', 'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
              'Wind Bearing (degrees)', 'Visibility (km)', 'Loud Cover', 'Pressure (millibars)']]
encode = pd.get_dummies(obj, drop_first = True)
df = pd.concat([other, encode], axis = 1)

In [342]:
df

Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0             9.472222                  7.388889      0.89            14.1197   
1             9.355556                  7.227778      0.86            14.2646   
2             9.377778                  9.377778      0.89             3.9284   
3             8.288889                  5.944444      0.83            14.1036   
4             8.755556                  6.977778      0.83            11.0446   
...                ...                       ...       ...                ...   
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
0                       251.0          15.8263         0.0   
1                       259.0          15.8263         0.0   
2                       204.0          14.9569         0.0   
3                       269.0          15.8263         0.0   
4                       259.0          15.8263         0.0   
...                       ...              ...         ...   
96448                    31.0          16.1000         0.0   
96449                    20.0          15.5526         0.0   
96450                    30.0          16.1000         0.0   
96451                    20.0          16.1000         0.0   
96452                    39.0          15.5204         0.0   

       Pressure (millibars)  Precip Type_rain  Precip Type_snow  
0                   1015.13                 1                 0  
1                   1015.63                 1                 0  
2                   1015.94                 1                 0  
3                   1016.41                 1                 0  
4                   1016.51                 1                 0  
...                     ...               ...               ...  
96448               1014.36                 1                 0  
96449               1015.16                 1                 0  
96450               1015.66                 1                 0  
96451               1015.95                 1                 0  
96452               1016.16                 1                 0  

[96453 rows x 10 columns]

In [343]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Temperature (C)           96453 non-null  float64
 1   Apparent Temperature (C)  96453 non-null  float64
 2   Humidity                  96453 non-null  float64
 3   Wind Speed (km/h)         96453 non-null  float64
 4   Wind Bearing (degrees)    96453 non-null  float64
 5   Visibility (km)           96453 non-null  float64
 6   Loud Cover                96453 non-null  float64
 7   Pressure (millibars)      96453 non-null  float64
 8   Precip Type_rain          96453 non-null  uint8  
 9   Precip Type_snow          96453 non-null  uint8  
dtypes: float64(8), uint8(2)
memory usage: 6.1 MB


## Feature scaling

In [344]:
df.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
count       96453.000000            96453.000000     96453.000000     96453.0   
mean           10.810640              187.509232        10.347325         0.0   
std             6.913571              107.383428         4.192123         0.0   
min             0.000000                0.000000         0.000000         0.0   
25%             5.828200              116.000000         8.339800         0.0   
50%             9.965900              180.000000        10.046400         0.0   
75%            14.135800              290.000000        14.812000         0.0   
max            63.852600              359.000000        16.100000         0.0   

       Pressure (millibars)  Precip Type_rain  Precip Type_snow  
count          96453.000000      96453.000000      96453.000000  
mean            1003.235956          0.883581          0.111059  
std              116.969906          0.320729          0.314207  
min                0.000000          0.000000          0.000000  
25%             1011.900000          1.000000          0.000000  
50%             1016.450000          1.000000          0.000000  
75%             1021.090000          1.000000          0.000000  
max             1046.380000          1.000000          1.000000

#### As we can see, Wind bearing and Pressure mean, std are very high tham Temperature. It will increase the error. Hence, scaling is necessary.

In [345]:
feature_scale = ['Wind Bearing (degrees)', 'Pressure (millibars)']
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
scaler.fit(df[feature_scale])

MinMaxScaler()

In [346]:
scaler.transform(df[feature_scale])

array([[0.69916435, 0.97013513],
       [0.72144847, 0.97061297],
       [0.56824513, 0.97090923],
       ...,
       [0.08356546, 0.97064164],
       [0.05571031, 0.97091879],
       [0.1086351 , 0.97111948]])

In [347]:
df1 = df.drop(['Wind Bearing (degrees)', 'Pressure (millibars)'], axis = 1)

In [348]:
final_data = pd.concat([df1.reset_index(drop=True), pd.DataFrame(scaler.transform(df[feature_scale]), columns=feature_scale)], axis = 1)

In [349]:
final_data

Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0             9.472222                  7.388889      0.89            14.1197   
1             9.355556                  7.227778      0.86            14.2646   
2             9.377778                  9.377778      0.89             3.9284   
3             8.288889                  5.944444      0.83            14.1036   
4             8.755556                  6.977778      0.83            11.0446   
...                ...                       ...       ...                ...   
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Visibility (km)  Loud Cover  Precip Type_rain  Precip Type_snow  \
0              15.8263         0.0                 1                 0   
1              15.8263         0.0                 1                 0   
2              14.9569         0.0                 1                 0   
3              15.8263         0.0                 1                 0   
4              15.8263         0.0                 1                 0   
...                ...         ...               ...               ...   
96448          16.1000         0.0                 1                 0   
96449          15.5526         0.0                 1                 0   
96450          16.1000         0.0                 1                 0   
96451          16.1000         0.0                 1                 0   
96452          15.5204         0.0                 1                 0   

       Wind Bearing (degrees)  Pressure (millibars)  
0                    0.699164              0.970135  
1                    0.721448              0.970613  
2                    0.568245              0.970909  
3                    0.749304              0.971358  
4                    0.721448              0.971454  
...                       ...                   ...  
96448                0.086351              0.969399  
96449                0.055710              0.970164  
96450                0.083565              0.970642  
96451                0.055710              0.970919  
96452                0.108635              0.971119  

[96453 rows x 10 columns]

## Modeling

### Linear regression

In [351]:
#separate dataset into train and test
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score
from sklearn import metrics

In [352]:
X = final_data.drop('Temperature (C)', axis = 1)
y = final_data['Temperature (C)']

In [353]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state = 0)

X_train.shape, X_test.shape

((67517, 9), (28936, 9))

In [354]:
regmodel = LinearRegression()

In [355]:
kfold_validation=KFold(10)

In [356]:
results=cross_val_score(regmodel, X_train, y_train, cv=kfold_validation)

In [357]:
regmodel.fit(X_train, y_train)

LinearRegression()

In [358]:
y_pred = regmodel.predict(X_test)

In [359]:
r_squared = r2_score(y_test, y_pred)

In [389]:
r_squared.round(2)

0.99

In [361]:
results

array([0.99025743, 0.99011504, 0.99022763, 0.99021398, 0.99032121,
       0.98997316, 0.99031961, 0.9900007 , 0.98989541, 0.98976527])

In [362]:
np.mean(results)

0.9901089443478108

In [391]:
print('Root mean squared error', np.sqrt(metrics.mean_squared_error(y_test, y_pred)).round(2))
print('Mean Absolute error: ', metrics.mean_absolute_error(y_test, y_pred).round(2))
print('Mean squared error: ', metrics.mean_squared_error(y_test,y_pred).round(2))

Root mean squared error 0.95
Mean Absolute error:  0.74
Mean squared error:  0.89


### Ridge regression

In [364]:
rreg = Ridge(alpha=.5)

In [365]:
rreg.fit(X_train, y_train)

Ridge(alpha=0.5)

In [366]:
y_pred2 = rreg.predict(X_test)

In [392]:
r2_score(y_test, y_pred2).round(2)

1.0

In [368]:
ridge_result = cross_val_score(rreg, X_train, y_train, cv=kfold_validation)

In [369]:
ridge_result

array([0.99025742, 0.99011504, 0.99022764, 0.99021398, 0.99032122,
       0.98997317, 0.99031959, 0.9900007 , 0.98989545, 0.98976525])

In [393]:
print('Root mean squared error', np.sqrt(metrics.mean_squared_error(y_test, y_pred2)).round(2))
print('Mean Absolute error: ', metrics.mean_absolute_error(y_test, y_pred2).round(2))
print('Mean squared error: ', metrics.mean_squared_error(y_test,y_pred2).round(2))

Root mean squared error 0.09
Mean Absolute error:  0.02
Mean squared error:  0.01


### Lasso regression

In [371]:
lreg = Lasso(alpha=.5)

In [372]:
lreg.fit(X_train, y_train)

Lasso(alpha=0.5)

In [373]:
y_pred3 = lreg.predict(X_test)

In [395]:
r2_score(y_test, y_pred3).round(2)

0.99

In [375]:
Lasso_result = cross_val_score(lreg, X_train, y_train, cv=kfold_validation)

In [376]:
Lasso_result

array([0.9896091 , 0.98948867, 0.98960297, 0.98960541, 0.98962801,
       0.98938578, 0.98963024, 0.98935833, 0.98933097, 0.98910124])

In [394]:
print('Root mean squared error', np.sqrt(metrics.mean_squared_error(y_test, y_pred3)).round(2))
print('Mean Absolute error: ', metrics.mean_absolute_error(y_test, y_pred3).round(2))
print('Mean squared error: ', metrics.mean_squared_error(y_test,y_pred3).round(2))

Root mean squared error 0.98
Mean Absolute error:  0.76
Mean squared error:  0.96


### Decision tree regression

In [378]:
from sklearn.tree import DecisionTreeRegressor

In [379]:
dregr = DecisionTreeRegressor()

In [380]:
dec_result = cross_val_score(dregr, X_train, y_train, cv=kfold_validation)

In [381]:
dregr.fit(X_train, y_train)

DecisionTreeRegressor()

In [382]:
y_pred2 = dregr.predict(X_test)

In [383]:
r2_score(y_test, y_pred2)

0.9999145599778249

In [384]:
dec_result

array([0.99988749, 0.99994516, 0.99992005, 0.99983487, 0.99992532,
       0.99995146, 0.99993032, 0.9999319 , 0.99988551, 0.99993693])

In [385]:
print('Root mean squared error', np.sqrt(metrics.mean_squared_error(y_test, y_pred2)))
print('Mean Absolute error: ', metrics.mean_absolute_error(y_test, y_pred2))
print('Mean squared error: ', metrics.mean_squared_error(y_test,y_pred2))

Root mean squared error 0.08824315151253367
Mean Absolute error:  0.023882399471633453
Mean squared error:  0.007786853788863975
